### ISSUE TITLE
node scoped computer can be considered out of scope while still provisoning

### ISSUE THREAD

**Issue by olivergondza** [jenkinsci/openstack-cloud-plugin#149]
A build failed with:
```
FATAL: java.io.IOException: SSH channel is closed
hudson.remoting.RequestAbortedException: java.io.IOException: SSH channel is closed
 at hudson.remoting.Request.abort(Request.java:297)
 at hudson.remoting.Channel.terminate(Channel.java:844)
 at hudson.remoting.Channel.close(Channel.java:1157)
 at hudson.remoting.Channel.close(Channel.java:1135)
 at jenkins.plugins.openstack.compute.JCloudsComputer.deleteSlave(JCloudsComputer.java:110)
 at jenkins.plugins.openstack.compute.JCloudsCleanupThread.deleteComputer(JCloudsCleanupThread.java:134)
 at jenkins.plugins.openstack.compute.JCloudsCleanupThread.terminatesNodesWithoutServers(JCloudsCleanupThread.java:182)
 at jenkins.plugins.openstack.compute.JCloudsCleanupThread.execute(JCloudsCleanupThread.java:66)
 at hudson.model.AsyncPeriodicWork$1.run(AsyncPeriodicWork.java:53)
 at java.lang.Thread.run(Thread.java:745)
 at ......remote call to ci-osp-rhel7-7631(Native Method)
 at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1413)
 at hudson.remoting.Request.call(Request.java:172)
 at hudson.remoting.Channel.call(Channel.java:777)
 at hudson.FilePath.act(FilePath.java:978)
 at hudson.FilePath.act(FilePath.java:967)
 at hudson.scm.SubversionSCM.checkout(SubversionSCM.java:894)
 at hudson.scm.SubversionSCM.checkout(SubversionSCM.java:830)
 at org.jenkinsci.plugins.multiplescms.MultiSCM.checkout(MultiSCM.java:129)
 at hudson.scm.SCM.checkout(SCM.java:484)
 at hudson.model.AbstractProject.checkout(AbstractProject.java:1274)
 at hudson.model.AbstractBuild$AbstractBuildExecution.defaultCheckout(AbstractBuild.java:609)
 at jenkins.scm.SCMCheckoutStrategy.checkout(SCMCheckoutStrategy.java:86)
 at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:531)
 at hudson.model.Run.execute(Run.java:1738)
 at hudson.matrix.MatrixRun.run(MatrixRun.java:146)
 at hudson.model.ResourceController.execute(ResourceController.java:98)
 at hudson.model.Executor.run(Executor.java:381)
Caused by: java.io.IOException: SSH channel is closed
 at com.trilead.ssh2.channel.ChannelManager.ioException(ChannelManager.java:1543)
 at com.trilead.ssh2.channel.ChannelManager.sendData(ChannelManager.java:373)
 at com.trilead.ssh2.channel.ChannelOutputStream.write(ChannelOutputStream.java:63)
 at com.trilead.ssh2.channel.ChannelOutputStream.write(ChannelOutputStream.java:68)
 at hudson.remoting.ChunkedOutputStream.sendFrame(ChunkedOutputStream.java:93)
 at hudson.remoting.ChunkedOutputStream.sendBreak(ChunkedOutputStream.java:66)
 at hudson.remoting.ChunkedCommandTransport.writeBlock(ChunkedCommandTransport.java:46)
 at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.write(AbstractSynchronousByteArrayCommandTransport.java:45)
 at hudson.remoting.Channel.send(Channel.java:579)
 at hudson.remoting.Channel.close(Channel.java:1153)
 at hudson.remoting.Channel.close(Channel.java:1135)
 at jenkins.plugins.openstack.compute.JCloudsComputer.deleteSlave(JCloudsComputer.java:110)
 at jenkins.plugins.openstack.compute.JCloudsCleanupThread.deleteComputer(JCloudsCleanupThread.java:134)
 at jenkins.plugins.openstack.compute.JCloudsCleanupThread.terminatesNodesWithoutServers(JCloudsCleanupThread.java:182)
 at jenkins.plugins.openstack.compute.JCloudsCleanupThread.execute(JCloudsCleanupThread.java:66)
 at hudson.model.AsyncPeriodicWork$1.run(AsyncPeriodicWork.java:53)
 at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: The connection is being shutdown
 at com.trilead.ssh2.channel.ChannelManager.handleEndMessage(ChannelManager.java:1525)
 at com.trilead.ssh2.transport.TransportManager$1.run(TransportManager.java:522)
 ... 1 more
Caused by: java.net.SocketException: No route to host
 at java.net.SocketInputStream.socketRead0(Native Method)
 at java.net.SocketInputStream.read(SocketInputStream.java:152)
 at java.net.SocketInputStream.read(SocketInputStream.java:122)
 at com.trilead.ssh2.crypto.cipher.CipherInputStream.fill_buffer(CipherInputStream.java:41)
 at com.trilead.ssh2.crypto.cipher.CipherInputStream.internal_read(CipherInputStream.java:52)
 at com.trilead.ssh2.crypto.cipher.CipherInputStream.getBlock(CipherInputStream.java:79)
 at com.trilead.ssh2.crypto.cipher.CipherInputStream.read(CipherInputStream.java:108)
 at com.trilead.ssh2.transport.TransportConnection.receiveMessage(TransportConnection.java:232)
 at com.trilead.ssh2.transport.TransportManager.receiveLoop(TransportManager.java:693)
 at com.trilead.ssh2.transport.TransportManager$1.run(TransportManager.java:489)
 ... 1 more
```
Here is the chain of events that caused it.

```
Apr 19, 2017 6:51:59 AM jenkins.plugins.openstack.compute.JCloudsSlaveTemplate provision
INFO: Provisioning new openstack server ci-osp-rhel7-7631 with options ...
...
Apr 19, 2017 6:52:32 AM jenkins.plugins.openstack.compute.JCloudsSlaveTemplate provision
INFO: Provisioned: NovaServer{id=XXX, name=ci-osp-rhel7-7631 ...
...
Apr 19, 2017 6:52:37 AM jenkins.plugins.openstack.compute.JCloudsCleanupThread destroyServersOutOfScope
INFO: Server ci-osp-rhel7-7631 run out of its scope node:ci-osp-rhel7-7631. Terminating.
...
Apr 19, 2017 6:55:21 AM jenkins.plugins.openstack.compute.JCloudsSlave createComputer
INFO: Creating a new computer for ci-osp-rhel7-7631
...
[04/19/17 06:55:27] SSH Launch of ci-osp-rhel7-7631 on 10.8.184.92 completed in 6,244 ms
...
Apr 19, 2017 6:56:26 AM hudson.node_monitors.AbstractAsyncNodeMonitorDescriptor monitor
WARNING: Failed to monitor ci-osp-rhel7-7631 for Architecture
java.util.concurrent.TimeoutException
...
Apr 19, 2017 7:03:10 AM jenkins.plugins.openstack.compute.JCloudsCleanupThread terminatesNodesWithoutServers
WARNING: No server running for computer ci-osp-rhel7-7631. Terminating.
Apr 19, 2017 7:03:10 AM jenkins.plugins.openstack.compute.JCloudsComputer deleteSlave
INFO: Deleting slave ci-osp-rhel7-7631
```

**Comment by olivergondza** [comment]
Hmm, the cause I have expected is fixed already: https://github.com/jenkinsci/openstack-cloud-plugin/commit/949f4060136f1324ad3b924a6404e961df3647eb

**Comment by olivergondza** [comment]
Releasing it a 2.22.

**Comment by slfletch** [comment]
I think I am getting the same/similar issue with version 2.22.  Do I need to manually update the openstack cloud plugin with your fix?  

`Checking label: ubuntu-16.04-slave

Jun 19, 2017 11:57:17 PM INFO hudson.model.AsyncPeriodicWork$1 run

Started OpenStack slave cleanup

Jun 19, 2017 11:57:17 PM WARNING jenkins.plugins.openstack.compute.ServerScope$Node isOutOfScope

No cloud-stats activity tracked for Ubuntu-16.04-slave-8726

Jun 19, 2017 11:57:17 PM INFO jenkins.plugins.openstack.compute.JCloudsCleanupThread destroyServersOutOfScope

Server Ubuntu-16.04-slave-8726 run out of its scope node:Ubuntu-16.04-slave-8726. Terminating: NovaServer{id=98ac730e-64f9-44cd-84d9-0e4f3ef9549a, name=Ubuntu-16.04-slave-8726, image={id=ae1f024a-0b34-4b58-b186-b55688417eb8, links=[{href=http://nova-api.openstack:8774/4a71466c464d4dbeac89e4ba8056da0f/images/ae1f024a-0b34-4b58-b186-b55688417eb8, rel=bookmark}]}, flavor=NovaFlavor{id=5185c458-37c3-439f-94a5-e48d435d61d0, ephemeral=0, swap=0, rxtx_factor=1.0, links=[GenericLink{href=http://nova-api.openstack:8774//flavors/5185c458-37c3-439f-94a5-e48d435d61d0, rel=bookmark}], 
}, status=ACTIVE, diskconfig=MANUAL, userId=, created=Mon Jun 19 23:51:59 UTC 2017, updated=Mon Jun 19 23:52:21 UTC 2017, launched at=Mon Jun 19 23:52:21 UTC 2017, tenantId=4a71466c464d4dbeac89e4ba8056da0f, hostId=xxx, addresses=NovaAddresses{addresses={private=[NovaAddress{address=xxx, type=fixed, version=4, macaddr=fa:xx, 
}, NovaAddress{address=xxx, type=floating, version=4, macaddr=fa:111, 
}]}, 
}, hypervisor host=cab24, powerstate=1, instanceName=instance-00000073, vmState=active, metadata={jenkins-scope=node:Ubuntu-16.04-slave-8726, jenkins-cloud-name=Openstack-cloud, jenkins-template-name=Ubuntu-16.04-slave, jenkins-instance=http://xxx/}}

Jun 19, 2017 11:57:18 PM INFO hudson.model.AsyncPeriodicWork$1 run

Finished OpenStack slave cleanup. 1,397 ms

Jun 19, 2017 11:57:24 PM WARNING org.eclipse.jetty.util.log.JavaUtilLog warn

Error while serving http://xxx/cloud/Openstack-cloud/provision
java.util.NoSuchElementException: No such server running: 98ac730e-64f9-44cd-84d9-0e4f3ef9549a
	at jenkins.plugins.openstack.compute.internal.Openstack.getServerById(Openstack.java:251)
	at jenkins.plugins.openstack.compute.JCloudsSlave.getPublicAddress(JCloudsSlave.java:115)
	at jenkins.plugins.openstack.compute.JCloudsCloud$SlaveType$1.isReady(JCloudsCloud.java:121)
	at jenkins.plugins.openstack.compute.JCloudsCloud.isSlaveReadyToLaunch(JCloudsCloud.java:356)
	at jenkins.plugins.openstack.compute.JCloudsSlaveTemplate.provisionSlave(JCloudsSlaveTemplate.java:187)
	at jenkins.plugins.openstack.compute.JCloudsCloud.doProvision(JCloudsCloud.java:435)
	at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
	at org.kohsuke.stapler.Function$MethodFunction.invoke(Function.java:343)
Caused: java.lang.reflect.InvocationTargetException`

**Comment by olivergondza** [comment]
@slfletch, I have [patched](https://github.com/jenkinsci/openstack-cloud-plugin/commit/6694936611c290ec4a468d3228dca316bfe5f530) this just yesterday and released as 2.23. Can you give it a try and reopen this if the problem is not gone? Thanks.

**Comment by slfletch** [comment]
Thanks! I plan on installing it tonight to try it out.
